[15.0][MIG] base_user_role_company

hbrunn opened the pull request:
this is a backport from v16 as it contains patches the previous v15 PRs (https://github.com/OCA/server-backend/pull/139, https://github.com/OCA/server-backend/pull/187) didn't

Chandresh-OSI committed:
[ADD] base_user_role_company: add per company user roles

oca-travis committed:
[UPD] Update base_user_role_company.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

kos94ok-3D committed:
[FIX] base_user_role_company: wrong xpath

OCA-git-bot committed:
base_user_role_company 14.0.1.1.0

Commit:
Added translation using Weblate (Italian)

dreispt committed:
[FIX] base_user_role_company: roles not properly applied on login

Commit:
Translated using Weblate (Italian)

Currently translated at 100.0% (11 of 11 strings)

Translation: server-backend-14.0/server-backend-14.0-base_user_role_company
Translate-URL: https://translation.odoo-community.org/projects/server-backend-14-0/server-backend-14-0-base_user_role_company/it/

dreispt committed:
[FIX] base_user_role_company: wrong menus on re-login

Issue found on logout / relogin.
The user groups were applied correctly, but the main menu showed apps
the user did not have access to.

This was related to the menu caching mechanisn, that was disabled here.

dreispt committed:
[FIX] base_user_role_company: fix tests

oca-travis committed:
[UPD] Update base_user_role_company.pot

OCA-git-bot committed:
base_user_role_company 14.0.2.0.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-backend-14.0/server-backend-14.0-base_user_role_company
Translate-URL: https://translation.odoo-community.org/projects/server-backend-14-0/server-backend-14-0-base_user_role_company/

OCA-git-bot committed:
base_user_role_company 14.0.2.0.1

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 90.9% (10 of 11 strings)

Translation: server-backend-14.0/server-backend-14.0-base_user_role_company
Translate-URL: https://translation.odoo-community.org/projects/server-backend-14-0/server-backend-14-0-base_user_role_company/it/

Urvisha-OSI committed:
[IMP] base_user_role_company: pre-commit stuff

Urvisha-OSI committed:
[MIG] base_user_role_company: Migration to 16.0

Commit:
[UPD] Update base_user_role_company.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-backend-16.0/server-backend-16.0-base_user_role_company
Translate-URL: https://translation.odoo-community.org/projects/server-backend-16-0/server-backend-16-0-base_user_role_company/

Ivorra78 committed:
Added translation using Weblate (Spanish)

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: server-backend-16.0/server-backend-16.0-base_user_role_company
Translate-URL: https://translation.odoo-community.org/projects/server-backend-16-0/server-backend-16-0-base_user_role_company/es/

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (8 of 8 strings)

Translation: server-backend-16.0/server-backend-16.0-base_user_role_company
Translate-URL: https://translation.odoo-community.org/projects/server-backend-16-0/server-backend-16-0-base_user_role_company/pt/

OCA-git-bot committed:
[UPD] README.rst

hbrunn committed:
[MIG] base_user_role_company

hbrunn commented on base_user_role_company/README.rst:
```diff
@@ -0,0 +1,119 @@
+Enable User Roles depending on the Companies selected.
+
+A company specific Role will only be enabled
+if it is set for **all** the currently selected companies.
```
@dreispt this isn't true since at least v14. To me it would make sense to adapt the code accordingly, or do you as maintainer say better adapt the readme?

JordiBForgeFlow replied:
@hbrunn  agree that the current README does not reflect what the code does.

The current code will poses a threat from a security standpoint. If you are assigned to Sales Manager role in company A only, you can still access to all sales manager functions in company B by just enabling this company together with company A.

I think that it's necessary to adapt the code to the README.

JordiBForgeFlow replied:
@hbrunn  agree that the current README does not reflect what the code does.

The current code will poses a threat from a security standpoint. If you are assigned to Sales Manager role in company A only, you can still access to all sales manager functions in company B by just enabling this company together with company A.

I think that it's necessary to adapt the code to the README.

sebalix replied:
Is there some news about this?

hbrunn replied:
fixed this right now

hbrunn replied:
@JordiBForgeFlow @sebalix @dreispt but wait, if this is supposed to be a security feature (as opposed to convenience as I saw it up to now), then the module needs to be revamped entirely:

We set the groups only on authentication or when get_session_info is called, that means if I authenticate, select company1 only, I'll have groups specific to company1.

For people just clicking things in the webclient, that's fine. But if I now craft a request for this session with a context setting allowed companies to company1,company2, this request runs with the groups specific to company1, but possibly on data of company2.

As we don't want to update groups for *every* request, and we can't rewrite `base_user_role` to not work by rewriting groups, I think we'll have to save somewhere server side for which companies the last update of groups has taken place, and restrict to/raise if we get allowed companies not in this list? With the current approach you can't be sensibly in company1 with one tab and company2 with another tab and the same user anyways, so this wouldn't really make things worse?

JordiBForgeFlow replied:
With the current data model, I don't see how to fix this issue. Seems as if we'd want to store, a model res.user.group.company that stores for which companies is a user assigned to a group..

hbrunn replied:
I think introducing a field `last_cids` on res.users that we write every time we rewrite groups should work. Then we can restrict the incoming allowed companies to the intersection with `last_cids`, and the company switcher then has to call a controller that does updating last_cids + group update in one transaction.

With this we should be able to guarantee you never have groups assigned for companies that shouldn't.

dgiraldoc2c replied:
Hello here!
I work for an Odoo integrator. I think this module could be quite useful for some clients. I subscribe to the analysis that it doesn't do what it's expected to as soon it is considered as a security module (which is how I see it and how I think it could be useful).
I'm a functional guy so I'm probably missing many technical aspects. Anyways, here is my take about what the module should do : 
- Regarding the Menus / Views / Access Rights, consider the intersection of what the companies ticked in the selector give access to (through the groups associated to the roles and their inherited groups). Example with Access Rights, if you have access to those two groups through companies ticked in the selector : 
![image](https://github.com/OCA/server-backend/assets/106684081/08be2db3-2ba3-4754-9d1e-9d55d067a523)
![image](https://github.com/OCA/server-backend/assets/106684081/b6eda770-16be-4b89-a57f-60a51bbe7f8a)
=> Then you should only get read and write access to sales orders and sale orders lines
- Regarding the record rules : apply an "AND" rule. In the example below we have 2 groups of record rules : 
![image](https://github.com/OCA/server-backend/assets/106684081/7c8f49f3-6f0d-4f66-af4c-ddff666e6ae4)
![image](https://github.com/OCA/server-backend/assets/106684081/d6271c3e-ac13-46e9-a96b-9eeb2fb3a4be)
=> In this example, since we have a record rule restricting the modification to one's own sales and the other to the sales in "draft" state (quotations), if the companies in the selector give access to those two groups, the modification should only be allowed for one's own sales in "draft" state.

What's your opinion ? Happy to discuss this topic with you !

hbrunn replied:
what you describe is changing how roles work in the first place. That would involve changing the base module, so very much out of scope here.

dgiraldoc2c replied:
Hello @hbrunn ,
Thanks for reading my post and coming back to me!
Do you have other solutions in mind ?

hbrunn replied:
what I wrote [above](https://github.com/OCA/server-backend/pull/241#discussion_r1495217717), but that also won't happen in this PR I think

hbrunn commented:
/ocabot migration base_user_role_company

JordiBForgeFlow commented:
@hbrunn Looks good. Perhaps a small glitch (not associated with the migration) is that when you access to the users associated to a role, from the role, the company assigned to the user is not visible.

![image](https://github.com/OCA/server-backend/assets/7683926/9d80a30c-d181-4fb2-81d7-cc616e096445)

It is only visible from the user:

![image](https://github.com/OCA/server-backend/assets/7683926/07596518-a870-4abc-b030-015953aae260)

hbrunn commented:
@JordiBForgeFlow I think I can slip this into this PR if nobody objects?

Do you have an opinion about https://github.com/OCA/server-backend/pull/241#pullrequestreview-1646067495 ?

JordiBForgeFlow commented:
@hbrunn I would add the change as a separate commit, so that we can fwp and backport.

hbrunn committed:
[IMP] base_user_role_company: show company in role too

hbrunn committed:
[FIX] base_user_role_company: actually only allow roles with companies

when this role is set for all currently active companies

hbrunn commented:
added the company column on the user tab in roles in https://github.com/OCA/server-backend/pull/241/commits/a2e03bf0f7f33c2741d87335d93009c155bbab4b, and fixed the behavior for multiple selected companies in https://github.com/OCA/server-backend/pull/241/commits/28c9cec393e1b8f430c28d096a0a393abffc1079

hbrunn committed:
fixup! [MIG] base_user_role_company
